services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/utils/ResumableFileDownloadSerializer.java

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.transfer.s3.internal.utils;
+
+import java.nio.file.Paths;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.Map;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.SdkField;
+import software.amazon.awssdk.core.protocol.MarshallLocation;
+import software.amazon.awssdk.core.protocol.MarshallingType;
+import software.amazon.awssdk.core.traits.LocationTrait;
+import software.amazon.awssdk.core.traits.TimestampFormatTrait;
+import software.amazon.awssdk.http.SdkHttpFullResponse;
+import software.amazon.awssdk.protocols.core.StringToInstant;
+import software.amazon.awssdk.protocols.core.StringToValueConverter;
+import software.amazon.awssdk.protocols.json.BaseAwsStructuredJsonFactory;
+import software.amazon.awssdk.protocols.json.StructuredJsonGenerator;
+import software.amazon.awssdk.protocols.json.internal.AwsStructuredPlainJsonFactory;
+import software.amazon.awssdk.protocols.json.internal.marshall.JsonMarshallerContext;
+import software.amazon.awssdk.protocols.json.internal.marshall.JsonMarshallerRegistry;
+import software.amazon.awssdk.protocols.json.internal.marshall.SimpleTypeJsonMarshaller;
+import software.amazon.awssdk.protocols.json.internal.unmarshall.JsonUnmarshaller;
+import software.amazon.awssdk.protocols.json.internal.unmarshall.JsonUnmarshallerContext;
+import software.amazon.awssdk.protocols.json.internal.unmarshall.JsonUnmarshallerRegistry;
+import software.amazon.awssdk.protocols.json.internal.unmarshall.SimpleTypeJsonUnmarshaller;
+import software.amazon.awssdk.protocols.jsoncore.JsonNode;
+import software.amazon.awssdk.protocols.jsoncore.JsonNodeParser;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.transfer.s3.DownloadFileRequest;
+import software.amazon.awssdk.transfer.s3.ResumableFileDownload;
+import software.amazon.awssdk.transfer.s3.S3TransferManager;
+import software.amazon.awssdk.utils.Logger;
+
+@SdkInternalApi
+public final class ResumableFileDownloadSerializer {
```

Comment:
Can we make this class flexible enough to support `ResumableFileUpload`?

Reply:
Lol, didn't see that one coming! Sure!

Reply:
On second thoughts, we don't know what fields `ResumableFileUpload` will have right now... Ignore my comment, we can refactor this class when we get to it.

Reply:
I extracted as much generic code as possible to a separate class, so a `ResumableFileUpload` serializer should be fairly quick to write.